### ISSUE TITLE
EL7 scan results experience

### ISSUE THREAD

**Issue by ewierschke** [plus3it/watchmaker#704]
### Intro:

I was looking for some feedback/confirmation/suggestions on the scan results of EL7 systems that have run watchmaker.  

My assumption is that watchmaker would account for potential shortcomings in scap content (from ash-linux) with a larger intent of meeting STIG requirements.

The scap scan results return a few failed settings as well as other third party scanners returning their own failed settings based on STIG requirements.  

I was hoping to either get some resolved by requesting updates be made to the ash-linux/watchmaker automation or potentially just get feedback on whether you have experience resolving inquries of the failed settings returned in scan results; with the added benefit of providing those details to other potential users of ash-linux/watchmaker.

<hr/>

### Main Findings:

From SCAP scan results on a RHEL7 build:

| My ref | Rule ID | Severity | Details | Notes | Action |
|------|------|------|------|------|------|
| <ul><li>- [x] A</li></ul> | Set LogLevel to INFO | low | LogLevel INFO should be set in /etc/ssh/sshd_config | found that line commented out in /etc/ssh/sshd_config | [SALTSTATE](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432350839) |
|<ul><li>- [x] B</li></ul>| Set SSH authentication attempt limit | medium | MaxAuthTries <num> should be set in /etc/ssh/sshd_config | found MaxAuthTries line with 6 commented out in /etc/ssh/sshd_config | [Test PAM Conflict](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432350839) |
|<ul><li>- [x] C</li></ul>| Set Boot Loader Password | high	| password should exist in grub.cfg | makes sense to not set for cloud-only systems | [SALTSTATE](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432353974) |
|<ul><li>- [x] D</li></ul>| Modify the System Login Banner| medium | banner exists via /etc/issue | unsure why result is fail | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432354608) |
|<ul><li>- [x] E</li></ul>| Ensure All SUID Executables Are Authorized | unknown | search for SUID files | unsure why result is fail | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432356719) |
|<ul><li>- [x] F</li></ul>| Add noexec Option to /dev/shm | unknown	| set noexec in /etc/fstab for /dev/shm | option is set in /etc/fstab but at end of list (check maybe looking for noexec on its own) | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432357896) |
|<ul><li>- [x] G</li></ul>| Add nodev Option to /tmp | unknown | nodev option should be set on /tmp in /etc/fstab | using tmpfs, therefore not /etc/fstab entry for /tmp? | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432360780) |
|<ul><li>- [x] H</li></ul>| Add noexec Option to /tmp | unknown | noexec option should be set on /tmp in /etc/fstab | using tmpfs for /tmp, therefore no specific /etc/fstab entry for /tmp | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432361168) |
|<ul><li>- [x] I</li></ul>| Add nosuid Option to /tmp | unknown | nosuid option should be set on /tmp in /etc/fstab | using tmpfs for /tmp, therefore no specific /etc/fstab entry for /tmp | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432361399) |
|<ul><li>- [x] J</li></ul>| Add nodev Option to /var/tmp | unknown | nodev option should be set on /var/tmp in /etc/fstab | is there a reason not to have a specific /etc/fstab entry for /var/tmp | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432363246) |
|<ul><li>- [x] K</li></ul>| Add noexec Option to /var/tmp | unknown	| noexec option should be set on /var/tmp in /etc/fstab | is there a reason not to have a specific /etc/fstab entry for /var/tmp | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432363567) |
|<ul><li>- [x] L</li></ul>| Add nosuid Option to /var/tmp | unknown	| nosuid option should be set on /var/tmp in /etc/fstab | no specific /etc/fstab entry for /var/tmp | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432363749) |
|<ul><li>- [x] M</li></ul>| Ensure auditd Collects Information on Kernel Module Loading and Unloading | medium | check provides guidance on configuring auditd rules | is there a way to add this to watchmaker | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432365821) |
|<ul><li>- [x] N</li></ul>| Record Attempts to Alter Logon and Logout Events | medium | check provides guidance on configuring auditd rules | is there a way to add this to watchmaker | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432365821) |
|<ul><li>- [x] O</li></ul>| Ensure auditd Collects Information on the Use of Privileged Commands | medium | check provides some guidance | is the only way to address this system/executable specific? | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432365821) |
|<ul><li>- [x] P</li></ul>| Ensure /var/tmp Located On Separate Partition | low | /var/tmp should be on a separate partition | curious why this isn't on its own partition | [DOCUMENT](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432369119) |

---

### Additional Findings

There are also certain EL7 STIG settings that don't appear to be getting set correctly from my understanding and would appreciate any details you might have.  It doesn't appear that the scap content or scan are checking for these correctly.

From third party scan results on CentOS 7 build:

| My ref | Rule ID | Severity | Details | Notes | Action |
|------|------|------|------|------|------|
| <ul><li>- [x] Q</li></ul> | [V-72243](https://www.stigviewer.com/stig/red_hat_enterprise_linux_7/2017-12-14/finding/V-72243) | Medium | The SSH daemon must not allow authentication using rhosts authentication. within `/etc/ssh/sshd_config` | `IgnoreRhosts yes` is commented out | [SALTSTATE](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432369505) |
| <ul><li>- [x] R</li></ul> | [V-72239](https://www.stigviewer.com/stig/red_hat_enterprise_linux_7/2017-12-14/finding/V-72239) | Medium | The SSH daemon must not allow authentication using RSA rhosts authentication. within `/etc/ssh/sshd_config` | `RhostsRSAAuthentication no` doesn't exist - no check in scap content, no entry in `/etc/ssh/sshd_config` | [WONTFIX](https://github.com/plus3it/watchmaker/issues/704#issuecomment-432603857) |

<hr/>

**Comment by ferricoxide** [comment]
Updated your comment so the table would actually render. Looking at the rendered content, nao

**Comment by ferricoxide** [comment]
I'm going to do these one-by-one:

Probably need to add a stand-alone handler for this:

A | Set LogLevel to INFO | low | LogLevel INFO should be set in /etc/ssh/sshd_config | found that line commented out in /etc/ssh/sshd_config
-- | -- | -- | -- | --

Regardless of whether it's STIG/SCAP prescribed, it's a Good To Have&trade;: if the connecting-user is presenting an SSH key for key-based authentication, that key's fingerprint will be logged so that audit-correlation can be done.

**Comment by ferricoxide** [comment]
This _should be_ superfluous:

B | Set SSH authentication attempt limit | medium | MaxAuthTries should be set in /etc/ssh/sshd_config | found MaxAuthTries line with 6 commented out in /etc/ssh/sshd_config
-- | -- | -- | -- | --

PAM is already configured to disallow more than three authentication-attempts before locking the account.

**Comment by ferricoxide** [comment]
This one's only practical for KickStart'ed (or similarly provisioned) or individually-managed systems:

C | Set Boot Loader Password | high | password should exist in grub.cfg | makes sense to not set for cloud-only systems
-- | -- | -- | -- | --

Basically, this is something that can never be put in a general-purpose project like watchmaker/ash/OSCAP: it would be beyond silly to set a bootlader password that's the same for every user of the automation-tool. 

That said, while one _could_ pull a value from a per-project wam config-file, it's silly to do so in contexts where interactive console access are not possible. It's potentially of value for physical systems or virtualization environments that provide unfettered, interactive (virtual) console access, but of dubious value to elsewhere. I'd probably do this as a (extremely) low-priority enhancement.

**Comment by ferricoxide** [comment]
System login banners are set. However, the associated validation pattern-matches are highly failure-prone. This should be a WONTFIX (especially given that the banners are provably present).

D | Modify the System Login Banner | medium | banner exists via /etc/issue | unsure why result is fail
-- | -- | -- | -- | --

**Comment by ferricoxide** [comment]
Some of these scan-items are essentially "never pass":

E | Ensure All SUID Executables Are Authorized | unknown | search for SUID files | unsure why result is fail
-- | -- | -- | -- | --

While one could argue that for RPM-managed files, a simple rpm-verify that file-modes match what's defined in the RPM, some STIG/SCAP remediations change modes ...and would miss SUID files that weren't RPM-managed any way.

Basically, the scan tool will alert on simple presence of SUID files to try to force a manual validation (presumably because of the above limitations)

**Comment by ferricoxide** [comment]
This is likely a scan-error:

F | Add noexec Option to /dev/shm | unknown | set noexec in /etc/fstab for /dev/shm | option is set in /etc/fstab but at end of list (check maybe looking for noexec on its own)
-- | -- | -- | -- | --

A check of a hardened system shows:

~~~~
$ grep -E /dev/shm.*noexec /etc/fstab
tmpfs           /dev/shm        tmpfs   rw,seclabel,nosuid,nodev,noexec 0 0
$ grep /dev/shm /proc/mounts | grep noexec
tmpfs /dev/shm tmpfs rw,seclabel,nosuid,nodev,noexec,relatime 0 0
~~~~

**Comment by ferricoxide** [comment]
OSCAP typically expects `/tmp` to be a "real" filesystem and isn't equipped to remediate. Similarly, most scan tools aren't equipped to properly scan when `/tmp` is a pseudofs:

G | Add nodev Option to /tmp | unknown | nodev option should be set on /tmp in /etc/fstab | using tmpfs, therefore not /etc/fstab entry for /tmp?
-- | -- | -- | -- | --

That said, the `/etc/systemd/system/tmp.mount.d/options.conf` file:

~~~~
$ sudo cat /etc/systemd/system/tmp.mount.d/options.conf
[Mount]
Options=mode=1777,strictatime,nosuid,noexec,nodev
~~~~

_does_ cause the pseudofs `/tmp` to be set up properly (and persistently):

~~~~
$ grep /tmp /proc/mounts
tmpfs /tmp tmpfs rw,seclabel,nosuid,nodev,noexec 0 0
~~~~

**Comment by ferricoxide** [comment]
See prior comment (re: `nodev` option):

H | Add noexec Option to /tmp | unknown | noexec option should be set on /tmp in /etc/fstab | using tmpfs for /tmp, therefore no specific /etc/fstab entry for /tmp
-- | -- | -- | -- | --

**Comment by ferricoxide** [comment]
See prior comment (re: nodev option):

I | Add nosuid Option to /tmp | unknown | nosuid option should be set on /tmp in /etc/fstab | using tmpfs for /tmp, therefore no specific /etc/fstab entry for /tmp
-- | -- | -- | -- | --

**Comment by ferricoxide** [comment]
J | Add nodev Option to /var/tmp | unknown | nodev option should be set on /var/tmp in /etc/fstab | is there a reason not to have a specific /etc/fstab entry for /var/tmp
-- | -- | -- | -- | --

`/var/tmp` is not a standalone filesystem:
* On systems where `/tmp` is a "real" filesystem, `/var/tmp` is supposed to be a bind-mount of `/tmp` (even though this violates the Linux prescriptions around the two directories/mounts). As such, it will inherit whatever mount-options are applied to `/tmp` (and fail if `/tmp` fails those same tests)
* On systems where `/tmp` is a pseudofs, `/var/tmp` is just a subdirectory of `/var` and will, therefor, be covered by the same mount-options as its parent filesystem.

In either case, it cannot have an entry in fstab that sets these values.

**Comment by ferricoxide** [comment]
Faulty scan:

M | Ensure auditd Collects Information on Kernel Module Loading and Unloading | medium | check provides guidance on configuring auditd rules | is there a way to add this to watchmaker
-- | -- | -- | -- | --

Already implemented in `/etc/sysctl.d/99-sysctl.conf`. However, most of the prescribed changes become invalid (won't show up in a `sysctl -p` output) after the other hardening rules have disabled IPv6.

**Comment by ferricoxide** [comment]
Likely a faulty scan:

N | Record Attempts to Alter Logon and Logout Events | medium | check provides guidance on configuring auditd rules | is there a way to add this to watchmaker
-- | -- | -- | -- | --

These are implemented in `/etc/audit/rules.d/perm_mod.rules` and `/etc/audit/audit.rules`.

**Comment by ferricoxide** [comment]
See prior note (re: "Logon and Logout Events"):

O | Ensure auditd Collects Information on the Use of Privileged Commands | medium | check provides some guidance | is the only way to address this system/executable specific?
-- | -- | -- | -- | --

**Comment by ferricoxide** [comment]
New requirement:

P | Ensure /var/tmp Located On Separate Partition | low | /var/tmp should be on a separate partition | curious why this isn't on its own partition
-- | -- | -- | -- | --

With `/var`, `/var/log/` and `/var/log/audit` already on their own partitions:
1) This would generally be "overkill" - yielding little protection vice the further wasting of resources it would cost to implement (i.e., needing to allocate yet another space-sucking filesystem that will never be "the right size" for a given tenants' use of a general-puropose system image)
2) On a system where `/var/tmp` is a bind-mount of `/tmp` (prescribed elsewhere - at least in earlier EL7 STIG versions), it is essentially already its own filesystem.

I'd probably mark this as either a WONTFIX or an _extremely_ low "enhancement". If tenants' accreditors demand this, they can extend the root VolumeGroup, create a new LVM of their preferred size and mount it at `/var/tmp`

**Comment by ferricoxide** [comment]
> From third party scan results on CentOS 7 build:
-- https://www.stigviewer.com/stig/red_hat_enterprise_linux_7/2017-12-14/finding/V-72243 - The SSH daemon must not allow authentication using rhosts authentication. - within '/etc/ssh/sshd_config' 'IgnoreRhosts yes' is commented out

This means that `IgnoreRhosts yes` is the default value.

**Comment by ferricoxide** [comment]
> -- https://www.stigviewer.com/stig/red_hat_enterprise_linux_7/2017-12-14/finding/V-72239 - The SSH daemon must not allow authentication using RSA rhosts authentication. - within '/etc/ssh/sshd_config' 'RhostsRSAAuthentication no' doesn't exist - no check in scap content, no entry in /etc/ssh/sshd_config

Option does not exist in the man page for `sshd_config(5)` or `sshd(8)`. This _usually_ means an option is not valid. If so, adding it to the `/etc/ssh/sshd_config` would likely result in the service failing to (re)start.

**Comment by lorengordon** [comment]
>> -- stigviewer.com/stig/red_hat_enterprise_linux_7/2017-12-14/finding/V-72239 - The SSH  daemon must not allow authentication using RSA rhosts authentication. - within '/etc/ssh/sshd_config' 'RhostsRSAAuthentication no' doesn't exist - no check in scap content, no entry in /etc/ssh/sshd_config

> Option does not exist in the man page for sshd_config(5) or sshd(8). This usually means an option is not valid. If so, adding it to the /etc/ssh/sshd_config would likely result in the service failing to (re)start.

The `RhostsRSAAuthentication` option seems to have been deprecated for EL7.4 and later. Make sure this third-party scanner is using correct definitions for the OS.

https://github.com/ComplianceAsCode/content/pull/2480
https://github.com/ComplianceAsCode/content/issues/2478

**Comment by ferricoxide** [comment]
@ewierschke [starting to work on](/ferricoxide/watchmaker/blob/Issue_704-Dox/docs/common_findings-EL7.md) creating documentation for the findings that we're _not_ going to alter functionality for. Can you review this document and let me know if you need greater specificity in the "replies"?

**Comment by lorengordon** [comment]
Closing as reported findings have been addressed either with custom salt states or documentation.
